client/executor/wasmtime/src/runtime.rs

```diff
@@ -158,42 +218,173 @@ directory = \"{cache_dir}\"
 	Ok(())
 }
 
+fn common_config() -> wasmtime::Config {
+	let mut config = wasmtime::Config::new();
+	config.cranelift_opt_level(wasmtime::OptLevel::SpeedAndSize);
+	config
+}
+
+pub struct Semantics {
+	/// Enabling this will lead to some optimization shenanigans that make calling [`WasmInstance`]
+	/// extermely fast.
+	///
+	/// Primarily this is achieved by not recreating the instance for each call and performing a
+	/// bare minimum clean up: reapplying the data segments and restoring the values for global
+	/// variables. The vast majority of the linear memory is not restored, meaning that effects
+	/// of previous executions on the same [`WasmInstance`] can be observed there.
+	///
+	/// This is not a problem for a standard substrate runtime execution because it's up to the
+	/// runtime itself to make sure that it doesn't involve any non-determinism.
+	///
+	/// Since this feature depends on instrumentation, it can be set only if [`CodeSupplyMode::Verbatim`]
+	/// is used.
+	pub fast_instance_reuse: bool,
+
+	/// The WebAssembly standard defines a call/value stack but it doesn't say anything about its
+	/// size except that it has to be finite. The implementations are free to choose their own notion
+	/// of limit: some may count the number of calls or values, others would rely on the host machine
+	/// stack and trap on reaching a guard page.
+	///
+	/// This obviously is a source of non-determinism during execution. This feature can be used
```

Comment:
First you say it is non deterministic and than you speak about it being able to be setup in a deterministic way?

Reply:
Yeah, that's right. The stack depth is not deterministic, but if you enable that feature (i.e. set `stack_depth_metering`) it will become deterministic.